packages/xpetra/src/CrsMatrix/Xpetra_TpetraCrsMatrix_decl.hpp

```diff
@@ -234,315 +180,229 @@ namespace Xpetra {
         const Teuchos::RCP<const Map<LocalOrdinal,GlobalOrdinal,Node> >& colMap,
         const Teuchos::RCP<const Map<LocalOrdinal,GlobalOrdinal,Node> >& domainMap = Teuchos::null,
         const Teuchos::RCP<const Map<LocalOrdinal,GlobalOrdinal,Node> >& rangeMap = Teuchos::null,
-        const Teuchos::RCP<Teuchos::ParameterList>& params = null)
-    : mtx_(Teuchos::rcp(new Tpetra::CrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>(lclMatrix, toTpetra(rowMap), toTpetra(colMap), toTpetra(domainMap), toTpetra(rangeMap), params))) {  }
+        const Teuchos::RCP<Teuchos::ParameterList>& params = null);
 #endif
 #endif
 
     //! Destructor.
-    virtual ~TpetraCrsMatrix() {  }
+    virtual ~TpetraCrsMatrix();
```

Comment:
If the destructor body is empty, you could (and should) just write this:
```c++
virtual ~TpetraCrsMatrix() = default;
```

Reply:
Looking through MueLu's library, I find many destructor weak symbols, e.g., `MueLu::Monitor::~Monitor()`.  I can get rid of these by declaring a nontrivial destructor:
```
// hpp
~Monitor();
// cpp
Monitor::~Monitor() {}
```

Four such destructors are about 3 MB in the MueLu library.

Question:  what are the downsides (if any) to doing this?

Reply:
Jonathan, isn't that a question about marking destructors virtual?

https://stackoverflow.com/questions/7403883/derived-class-with-non-virtual-destructor


Just doing:
```
// hpp
~Monitor() = default;
```
 Should be the same as ` defining w/ {}`

Reply:
@jjellio It appears they are not the same.

If I do
```
// hpp
~FactoryMonitor() = default;
```
then I see many symbols in `libmuelu.a` of the form
```
W MueLu::FactoryMonitor::~FactoryMonitor()
```

If I do
```
// hpp
~FactoryMonitor();
// cpp
FactoryMonitor::~FactoryMonitor();
```
then in `libmuelu.a`, I see only
```
U MueLu::FactoryMonitor::~FactoryMonitor()
```

More importantly, in the latter case, the library is 1.1 MB smaller.

Reply:
Same as what we talked about  in email.

```
~FactoryMonitor() = default;
```
is placing the definition within the class decl, so it is inlined.

```
FactoryMonitor::~FactoryMonitor();
```

Declares the destructor, but you better provide a def in the def file. (I don't think you can define it using `default`, the only way would be to do `FactoryMonitor::~FactoryMonitor() {}` in the def, unless I am wrong)

Reply:
Btw, you could try:

in Foo_def.hpp
```
FactoryMonitor::~FactoryMonitor() = default;
```
Foo_decl.hpp
```
~FactoryMonitor();
```

Reply:
Ah, you are right.  However, back to my question.  Why would I choose to use the default ctor, which appears multiple times as a weak symbol in the library and thus increases the library's size, instead of placing it in the definition?

Reply:
default is just sugar for `{}` as I understand it.

What if you put in the def file,
```
FactoryMonitor::~FactoryMonitor() = default;
```

Reply:
I just don't know if you can use 'default' outside a class declaration ... you could always write
Foo_def.hpp
```
FactoryMonitor::~FactoryMonitor() = {}
```

Reply:
Oh, and since this is a Base Class, you I think you have to define the destructor virtual.

Reply:
> Oh, and since this is a Base Class, you I think you have to define the destructor virtual.

Yep.